tests/integration/test_etcd.py

```diff
@@ -0,0 +1,206 @@
+    assert "etcd" in results
+    for key in ["prefix", "tls-ca", "username", "version", "endpoints"]:
+        assert key in results["etcd"]
+    assert results["etcd"]["username"] == "test-common-name"
+
+    # check if secrets are used on Juju3
+    assert await check_secrets_usage_matching_juju_version(
+        ops_test,
+        ops_test.model.applications[DATA_INTEGRATOR].units[0].name,
+        integrator_relation.id,
+    )
+
+
+@pytest.mark.group(1)
+@only_on_localhost
+@only_with_juju_secrets
+@pytest.mark.abort_on_fail
+async def test_read_write(ops_test: OpsTest, cloud_name: str):
+    """Write and read to the key prefix with the requirer charm."""
+    # download etcdctl binary
+    urlretrieve(
+        "https://github.com/etcd-io/etcd/releases/download/v3.5.18/etcd-v3.5.18-linux-amd64.tar.gz",
+        "etcd-v3.5.18-linux-amd64.tar.gz",
+    )
+    # extract etcdctl binary
+    with tarfile.open("etcd-v3.5.18-linux-amd64.tar.gz", "r:gz") as tar:
+        tar.extractall()
+    Path("etcd-v3.5.18-linux-amd64/etcdctl").rename("etcdctl")
+
+    # get endpoints
+    # run get credentials action on data-integrator
+    action = (
+        await ops_test.model.applications[DATA_INTEGRATOR].units[0].run_action("get-credentials")
+    )
+    action = await action.wait()
+    assert action.status == "completed"
+    endpoints = action.results["etcd"]["endpoints"]
+
+    # write server certificate to disk
+    Path("ca.pem").write_text(
+        action.results["etcd"]["tls-ca"],
+    )
+
+    # get certificate from config of data-integrator
+    result = subprocess.check_output(
+        [
+            "./etcdctl",
+            "--endpoints",
+            endpoints,
+            "--cert",
+            "./client.pem",
+            "--key",
+            "./client.key",
+            "--cacert",
+            "./ca.pem",
+            "put",
+            "/test/test-key",
+            "test-value",
+        ],
+    ).strip()
+    assert result.decode().strip() == "OK"
+
+    # read the key
+    result = subprocess.check_output(
+        [
+            "./etcdctl",
+            "--endpoints",
+            endpoints,
+            "--cert",
+            "./client.pem",
+            "--key",
+            "./client.key",
+            "--cacert",
+            "./ca.pem",
+            "get",
+            "/test/test-key",
+        ],
+    ).strip()
+    assert result.decode().strip().split("\n") == ["/test/test-key", "test-value"]
+
+
+@pytest.mark.group(1)
+@only_on_localhost
+@only_with_juju_secrets
+@pytest.mark.abort_on_fail
+async def test_clean(ops_test: OpsTest, cloud_name: str):
+    # delete files used in tests
+    Path("client.pem").unlink(missing_ok=True)
+    Path("client.key").unlink(missing_ok=True)
+    Path("ca.pem").unlink(missing_ok=True)
+    Path("etcd-v3.5.18-linux-amd64.tar.gz").unlink(missing_ok=True)
+    Path("etcdctl").unlink(missing_ok=True)
+    rmtree("etcd-v3.5.18-linux-amd64", ignore_errors=True)
```

Comment:
We should most probably move the `etcdctl` install in the test charm. That way we won't need to persist and clean up anything in the test itself.